### PR TITLE
[com_ajax] to support templates

### PR DESCRIPTION
**Pull request by bassmanpaul**
See [this discussion](https://github.com/joomla/joomla-cms/issues/16382#issue-232511097) for why template support is needed in `com_ajax`.

I have basically copied the module logic which requires a helper file in the root of the extension with specific class & method names.

See the below code to quickly test support for this from the front end & the back end.

#### templates/protostar/helper.php:
```php
class tplProtostarHelper
{
	public static function getAjax()
	{
		return __METHOD__ . "\n\n" . JText::_( 'TPL_PROTOSTAR_XML_DESCRIPTION' );
	}

	public static function mySuperAwesomeMethodToTriggerAjax()
	{
		return __METHOD__ . "\n\n" . JText::_( 'TPL_PROTOSTAR_XML_DESCRIPTION' );
	}
}
```

#### Ajax call test:

```javascript
// Default method:
jQuery.ajax( 'index.php?option=com_ajax&template=protostar&format=json' );

// Custom method:
jQuery.ajax( 'index.php?option=com_ajax&template=protostar&format=json&method=mySuperAwesomeMethodToTrigger' );
```
#### Expected results:

- When called from the front end in protostar template: ✔ success.
- When called from the front end in another template: ✖ unauthorised.
- When called from the back end in any template: ✔ success.

#### templates/hathor/helper.php:
```php
class tplHathorHelper
{
	public static function getAjax()
	{
		return __METHOD__ . "\n\n" . JText::_( 'TPL_HATHOR_XML_DESCRIPTION' );
	}

	public static function mySuperAwesomeMethodToTriggerAjax()
	{
		return __METHOD__ . "\n\n" . JText::_( 'TPL_HATHOR_XML_DESCRIPTION' );
	}
}
```

#### Ajax call test:

```javascript
// Default method:
jQuery.ajax( 'index.php?option=com_ajax&template=hathor&format=json' );

// Custom method:
jQuery.ajax( 'index.php?option=com_ajax&template=hathor&format=json&method=mySuperAwesomeMethodToTrigger' );
```

#### Expected results:

- When called from the front end in any template: ✖ unauthorised.
- When called from the back end in any template: ✔ success.

### CODE REVIEW

File: components/com_ajax/ajax.php [hunk 1]
```diff
@@ -148,6 +148,99 @@
...
+ */
+elseif ($input->get('template'))
+{
+	$template = $input->get('template');
+
+	/*
+	 * Only allow access to current menu item's template (like modules) in the front end.
+	 * Allow full access in the back end as user is authenticated.
+	 */
+	if (($app->isSite() && $template === $app->getTemplate()) || $app->isAdmin())
```
Comment (laoneo) [inline components/com_ajax/ajax.php]: Use the isClient function please

Comment (bassmanpaul) [reply components/com_ajax/ajax.php]: Done :)

**Comment by bassmanpaul** [comment]
When then is implemented the documentation will also need updating:

https://docs.joomla.org/Using_Joomla_Ajax_Interface

**Comment by Bakual** [comment]
I wouldn't restrict it to the template of the current menuitem. I don't see a reason for that restriction. For modules we also only check for the published state and not if it's actually visible.
I would just check if the template is published.

**Comment by bassmanpaul** [comment]
Done.

Is retrieving the template published state best accessed through `JTable::getInstance('extension');`?

**Comment by Bakual** [comment]
Since we don't seem to have a JTemplateHelper::isEnabled method (like we have for components, plugins and modules), I would probably just use a direct database query. But using the table should work as well.

**Comment by bassmanpaul** [comment]
Yeah, I did wonder whether it would be helpful to suggest creating a Template Helper as modules, plugins & languages all have their own helpers & classes.

I've used:

```php
$table		= JTable::getInstance('extension');
$templateid	= $table->find(array('type' => 'template', 'element' => $template));

if ($templateid && $table->load($templateid) && $table->enabled)
...
```

**Comment by Bakual** [comment]
For consistency it would make sense, yes. Not sure how useful it will be.

File: components/com_ajax/ajax.php [hunk 2]
```diff
@@ -148,6 +148,97 @@
...
+ * Template support.
+ *
+ * tplFooHelper::getAjax() is called where 'foo' is the value
+ * of the 'template' variable passed via the URL
+ * (i.e. index.php?option=com_ajax&template=foo).
+ *
+ */
+elseif ($input->get('template'))
+{
+	$template	= $input->get('template');
```
Comment (matrikular) [inline components/com_ajax/ajax.php]: Could you remove the unnecessary whitespace, please? If you really need alignment, it would be one space after the longest _camelCased_ variable name.

File: components/com_ajax/ajax.php [hunk 3]
```diff
@@ -148,6 +148,97 @@
...
+		}
+		else
+		{
+			$class = 'Tpl' . ucfirst($template) . 'Helper';
+		}
+
+		$method = $input->get('method') ?: 'get';
+
+		// Is the back end accessing a front end template?
+		if ($app->isClient('administrator') && !is_file($helperFile))
```
Comment (matrikular) [inline components/com_ajax/ajax.php]: If I understand this check correctly, for this to work (even as a fallback), I would have to have two templates (frontend / backend) installed with the same name, right?

I would say that we should not look for a fallback helper in the opposite context if one wasn't found.

File: components/com_ajax/ajax.php (hunk 2 above)
Comment (bassmanpaul) [reply components/com_ajax/ajax.php]: To clarify, are you asking for the following:

```php
	$template = $input->get('template');
	$table = JTable::getInstance('extension');
	$templateid = $table->find(array('type' => 'template', 'element' => $template));
```

As I tried to follow the previous convention in the above plugin code block by lining up the equals:

```php
	$group      = $input->get('group', 'ajax');
	JPluginHelper::importPlugin($group);
	$plugin     = ucfirst($input->get('plugin'));
	$dispatcher = JEventDispatcher::getInstance();
```

File: components/com_ajax/ajax.php (hunk 3 above)
Comment (bassmanpaul) [reply components/com_ajax/ajax.php]: Basically, when in the back-end **template style config page** (for example editing `protostar`'s configuration) you want to call the `protostar` template's helper file, but because you are in the back-end `JPATH_BASE` will look in the administrator's templates which would be incorrect. Therefore, as you hinted, this is a fallback to allow for retrieving a front-end template from the back-end. However, alternative suggestions are welcome :)

As you mentioned; If you have a back-end template called `foo` and are in the back-end requesting your front-end template which is also called `foo` then the back-end one will be called. So this would be one caveat to be aware of.

File: components/com_ajax/ajax.php (hunk 2 above)
Comment (matrikular) [reply components/com_ajax/ajax.php]: Something like this:
```
$template   = $input->get('template');
$table      = JTable::getInstance('extension');
$templateId = $table->find(array('type' => 'template', 'element' => $template));
```

File: components/com_ajax/ajax.php (hunk 3 above)
Comment (matrikular) [reply components/com_ajax/ajax.php]: My mistake. By just looking at the code I somehow forgot the purpose of this PR. How about building the path to the helper file based on the client_id?

File: components/com_ajax/ajax.php (hunk 2 above)
Comment (bassmanpaul) [reply components/com_ajax/ajax.php]: Sure, updated.

File: components/com_ajax/ajax.php (hunk 3 above)
Comment (bassmanpaul) [reply components/com_ajax/ajax.php]: That would just be clever :) 

How about setting the `$basePath` that controls everything like:

```php
$basePath = ($table->client_id) ? JPATH_ROOT . '/administrator' : JPATH_ROOT;
```

Then remove the `$app->isClient('administrator')` logic. This way `client_id` would equal **1** for the admin side and set the root path.

Comment (Bakual) [reply components/com_ajax/ajax.php]: Sounds good to me

Comment (bassmanpaul) [reply components/com_ajax/ajax.php]: Done.

Would it also be worth allowing the user to also specify the `client_id` in the ajax url, to account for back-end/front-end template name conflicts as @matrikular mentioned. Or is the scenario too rare to warrant accounting for?

File: components/com_ajax/ajax.php [hunk 4]
```diff
@@ -148,6 +148,90 @@
...
+ * (i.e. index.php?option=com_ajax&template=foo).
+ *
+ */
+elseif ($input->get('template'))
+{
+	$template   = $input->get('template');
+	$table      = JTable::getInstance('extension');
+	$templateId = $table->find(array('type' => 'template', 'element' => $template));
+
+	if ($templateid && $table->load($templateid) && $table->enabled)
```
Comment (matrikular) [inline components/com_ajax/ajax.php]: Notice: Undefined variable: template**i**d instead of template**I**d.

Comment (bassmanpaul) [reply components/com_ajax/ajax.php]: Sorted, thanks.

File: components/com_ajax/ajax.php (hunk 3 above)
Comment (matrikular) [reply components/com_ajax/ajax.php]: I'll be happy to implement and / or test such an addition if one finds him/herself in that dilemma anytime soon but would leave it as it is for now.

**Comment by matrikular** [comment]
I have tested this item :white_check_mark: successfully on 8bbf94d046e5500d114c4852ecec84d2cbcbc542<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16532">issues.joomla.org/tracker/joomla-cms/16532</a>.</sub>

**Comment by bassmanpaul** [comment]
What's the next step for this feature? Do I need to run more tests or is it pending the next Joomla release?

**Comment by Bakual** [comment]
Usually, we want two people testing a PR. Your own test will not count as we assume you tested your own code prior to submitting it :smile: 

When those tests are done, it's the call of the release leader to include it into the next release. earliest possible release will be 3.8.0 since it is a new feature.

**Comment by bassmanpaul** [comment]
Okay, sure. Thanks for letting me know :)

**Comment by grig0ry** [comment]
I have tested this item :white_check_mark: successfully on 8bbf94d046e5500d114c4852ecec84d2cbcbc542<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16532">issues.joomla.org/tracker/joomla-cms/16532</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

**Comment by zero-24** [comment]
@mbabker do you want to have this in 3.8?

**Comment by grig0ry** [comment]
@zero-24 It would be great.

**Comment by zero-24** [comment]
> It would be great.

@Grigory90 I'm not the release leader so I can't take the decision on that. Lets give Michael a bit time to decide on that.

File: components/com_ajax/ajax.php [hunk 5]
```diff
@@ -148,6 +148,90 @@
...
+ */
+elseif ($input->get('template'))
+{
+	$template   = $input->get('template');
+	$table      = JTable::getInstance('extension');
+	$templateId = $table->find(array('type' => 'template', 'element' => $template));
+
+	if ($templateId && $table->load($templateId) && $table->enabled)
+	{
+		$basePath   = ($table->client_id) ? JPATH_ROOT . '/administrator' : JPATH_ROOT;
```
Comment (mbabker) [inline components/com_ajax/ajax.php]: Use `JPATH_ADMINISTRATOR` and `JPATH_SITE` path constants.
